### PR TITLE
add endpoint to search spaces by creator

### PR DESCRIPTION
**Pull request by mattcasey**
### <samp>🤖 Generated by Copilot at 6819948</samp>

This pull request adds a new feature to the public API that allows searching for spaces by the creator's wallet address. It updates the `Space` type and schema, refactors the `createWorkspaceApi` function, and adds a new module `lib/public-api/searchSpaces` and a new route `pages/api/v1/spaces/search` to implement the feature. It also adds the corresponding documentation and tests.

### WHY
We have no visibility into how many requests are being made to our public api, or which endpoints are being used and by whom. This should solve that

### <samp>🤖 Generated by Copilot at 6819948</samp>

*  Add a new feature to the public API that allows searching for spaces by the creator's wallet address ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-832ad9576589a1b3f65bd2f6990d503d84814ed38fdfa9d2206d0ac919415219R1-R29), [link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-94d0b42c2e0286a3da9708574ea7b00fb0f0880bbc66474751a0a59915b673e2R1-R55))
  * Create a new module `lib/public-api/searchSpaces` that exports a function `searchSpaces` and a type `SearchSpacesInput` ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-832ad9576589a1b3f65bd2f6990d503d84814ed38fdfa9d2206d0ac919415219R1-R29))
  * Create a new API route `pages/api/v1/spaces/search` that handles GET requests for searching spaces ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-94d0b42c2e0286a3da9708574ea7b00fb0f0880bbc66474751a0a59915b673e2R1-R55))
* Refactor the `createWorkspaceApi` function to use a helper function `mapSpace` that converts a space object from the database to a `Space` type ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-1d75640da4cedec692e7490b000d30de6ae9710b17b2e40067b15d3d5565c3f7R10), [link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-1d75640da4cedec692e7490b000d30de6ae9710b17b2e40067b15d3d5565c3f7L102-R112))
  * Import the `Space` type from the `lib/public-api/interfaces` module ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-1d75640da4cedec692e7490b000d30de6ae9710b17b2e40067b15d3d5565c3f7R10))
* Split the `CreateWorkspaceResponseBody` interface into two interfaces: `Space` and `CreateWorkspaceResponseBody` ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-75383c8ec6121cabaa880e2e0eed60278124a78ec22cc4ea2fc15d878a09557bL353-R391))
* Rename the `CreateWorkspaceResponseBody` schema to `Space` in the Swagger documentation ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-75383c8ec6121cabaa880e2e0eed60278124a78ec22cc4ea2fc15d878a09557bL339-R339))
* Add the `avatar` and `createdAt` properties to the `Space` schema in the Swagger documentation ([link](https://github.com/charmverse/app.charmverse.io/pull/2091/files?diff=unified&w=0#diff-75383c8ec6121cabaa880e2e0eed60278124a78ec22cc4ea2fc15d878a09557bR346-R354))

### CODE REVIEW

File: pages/api/v1/spaces/search.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import { superApiHandler } from 'lib/public-api/handler';
+import type { Space } from 'lib/public-api/interfaces';
+import type { SearchSpacesInput } from 'lib/public-api/searchSpaces';
+import { searchSpaces } from 'lib/public-api/searchSpaces';
+
+export const handler = superApiHandler().get(search);
+
+/**
+ * @swagger
+ * /users/search:
```
Comment (Devorein) [inline pages/api/v1/spaces/search.ts:13]: The docs here seem to be inaccurate. It should be `space/search`

File: pages/api/v1/spaces/search.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+async function search(req: NextApiRequest, res: NextApiResponse<Space[]>) {
+  const request = req.query as Omit<SearchSpacesInput, 'apiKey'>;
+
+  if (!request.userWallet) {
+    throw new InvalidStateError('userWallet is required');
+  }
+
+  const result = await searchSpaces(request);
+
+  const allowedResult = result.filter((space) => req.spaceIdRange?.includes(space.id));
```
Comment (Devorein) [inline pages/api/v1/spaces/search.ts:44]: If there is `spaceIdRange` in the payload why don't we use it while searching the space?

**Review by Devorein — APPROVED** [review]

File: pages/api/v1/spaces/search.ts (hunk 2 above)
Comment (mattcasey) [reply pages/api/v1/spaces/search.ts:44]: We realized we needed this after i wrote searchSpaces(), but I decided to leave the 'authorization' layer of logic in the middleware instead. Feels cleaner and we might want to use searchSpaces for something else.

File: pages/api/v1/spaces/search.ts (hunk 1 above)
Comment (mattcasey) [reply pages/api/v1/spaces/search.ts:13]: thanks!

**Pull request merged**